src/librbd/io/AioCompletion.cc

```diff
@@ -262,18 +257,14 @@ void AioCompletion::complete_external_callback() {
 }
 
 void AioCompletion::mark_complete_and_notify() {
-  state = AIO_STATE_COMPLETE;
+  completed.store(true, std::memory_order_release);
+  completed.notify_all();
```

Comment:
I'd place this call where the corresponding call on `cond` was.  The event socket has always been notified first, so unless there is a good rationale for the change it's better to keep it that way:
```
    ictx->event_socket.notify();
  }

  completed.notify_all();

  if (image_dispatcher_ctx != nullptr) {
```